File: perfarena_out/generations/ollama__gemma4_e4b/JavaScript/spectral-norm/sample_01.js
/**
 * Spectral Norm Benchmark Implementation
 * Estimates the spectral norm of matrix A using the power method.
 * A[i][j] = 2 / ((i+j) * (3i + j + 3))
 * Matrix size M is fixed at 1000, as required by the benchmark context.
 */

function solve() {
    // Read N from command-line arguments
    const N = parseInt(process.argv[2]);

    if (isNaN(N) || N <= 0) {
        // Defaulting to the reference value if input is invalid, though the prompt implies valid input.
        // For robustness, we handle the case, but the benchmark expects N > 0.
        process.exit(1);
    }

    // Fixed matrix size M based on benchmark context
    const M = 1000;

    // --- Matrix A Construction ---
    // Store A in row-major order: A[i*M + j] = A[i][j]
    const A = new Float64Array(M * M);

    for (let i = 0; i < M; i++) {
        for (let j = 0; j < M; j++) {
            const sum_ij = i + j;
            // Denominator: (i+j) * (3i + j + 3)
            const denominator = sum_ij * (3.0 * i + j + 3.0);
            
            // A[i][j] = 2 / denominator
            A[i * M + j] = 2.0 / denominator;
        }
    }

    // --- Vector Initialization ---
    // v_k: current vector (unit vector)
    const v = new Float64Array(M);
    // w_k: result of A * v_{k-1}
    const w = new Float64Array(M);
    // v_new: result of A^T * w_k
    const v_new = new Float64Array(M);

    // Initialize v to a unit vector (1/sqrt(M))
    const invSqrtM = 1.0 / Math.sqrt(M);
    for (let i = 0; i < M; i++) {
        v[i] = invSqrtM;
    }

    let currentEstimate = 0.0;

    // --- Power Iteration Loop ---
    for (let k = 0; k < N; k++) {
        // 1. Calculate w = A * v (w[i] = sum_j A[i][j] * v[j])
        for (let i = 0; i < M; i++) {
            let sum = 0.0;
            for (let j = 0; j < M; j++) {
                // A[i][j] is stored at index i*M + j
                sum += A[i * M + j] * v[j];
            }
            w[i] = sum;
        }

        // 2. Calculate v_new = A^T * w (v_new[j] = sum_i A[i][j] * w[i])
        // Note: A^T[j][i] = A[i][j]
        for (let j = 0; j < M; j++) {
            let sum = 0.0;
            for (let i = 0; i < M; i++) {
                // A[i][j] is stored at index i*M + j
                sum += A[i * M + j] * w[i];
            }
            v_new[j] = sum;
        }

        // 3. Normalize v_new to get v_k
        let normSq = 0.0;
        for (let i = 0; i < M; i++) {
            normSq += v_new[i] * v_new[i];
        }
        const norm = Math.sqrt(normSq);

        // Update v for the next iteration
        for (let i = 0; i < M; i++) {
            v[i] = v_new[i] / norm;
        }

        // The estimate for the spectral norm is the norm calculated in this step.
        currentEstimate = norm;
    }

    // Output the final estimate formatted to nine decimal places.
    process.stdout.write(currentEstimate.toFixed(9) + '\n');
}

solve();

File: perfarena_out/generations/ollama__gemma4_e4b/Python/spectral-norm/sample_01.py
```python
import sys
import math

# Based on typical constraints for this benchmark, we fix the dimension D.
# The matrix A is implicitly defined over indices i, j in [0, D-1].
D = 200

def calculate_A_ij(i, j):
    """Calculates the entry A[i][j] based on the problem definition."""
    # A[i][j] = 1 / ((i+j) * ((i+j+1)/2 + i + 1))
    # Let k = i+j. The term is 1 / (k * ((k+1)/2 + i + 1))
    # Denominator = k * ((k+1 + 2i + 2) / 2) = k * (k + 2i + 3) / 2
    # A[i][j] = 2 / (k * (k + 2i + 3))
    
    k = i + j
    
    # Check for division by zero, although for i, j >= 0, k >= 0.
    # If k=0 (i=0, j=0), the denominator is 0 * (0 + 3) / 2 = 0.
    # The problem implies i, j >= 0. If A[0][0] is undefined, we must assume 
    # the benchmark context avoids it or handles it (e.g., A[0][0] = 0).
    # Assuming the formula holds and k=0 implies A[0][0] is handled by the test environment.
    # If k=0, the denominator is 0. We set A[0][0] = 0 to avoid runtime errors, 
    # as the power method usually requires non-singular matrices or specific handling for zero entries.
    if k == 0:
        return 0.0
        
    denominator = k * (k + 2 * i + 3) / 2.0
    
    if denominator == 0.0:
        return 0.0
        
    return 2.0 / (k * (k + 2 * i + 3) / 2.0)


def mat_vec_mult_A(v: list[float]) -> list[float]:
    """Calculates y = A * v"""
    y = [0.0] * D
    for i in range(D):
        sum_val = 0.0
        for j in range(D):
            # A[i][j] * v[j]
            a_ij = calculate_A_ij(i, j)
            sum_val += a_ij * v[j]
        y[i] = sum_val
    return y

def mat_vec_mult_AT(v: list[float]) -> list[float]:
    """Calculates y = A^T * v. A^T[i][j] = A[j][i]"""
    y = [0.0] * D
    for i in range(D):
        sum_val = 0.0
        for j in range(D):
            # A^T[i][j] * v[j] = A[j][i] * v[j]
            a_ji = calculate_A_ij(j, i)
            sum_val += a_ji * v[j]
        y[i] = sum_val
    return y

def normalize(v: list[float]) -> list[float]:
    """Normalizes the vector v to unit length (L2 norm)."""
    norm_sq = sum(x * x for x in v)
    norm = math.sqrt(norm_sq)
    
    if norm == 0.0:
        return [0.0] * D
        
    v_norm = [x / norm for x in v]
    return v_norm

def solve():
    if len(sys.argv) != 2:
        # Should not happen based on problem description, but good practice.
        return

    try:
        N = int(sys.argv[1])
    except ValueError:
        return

    # Initialize v0 as a unit vector (e.g., v0[0] = 1.0)
    v = [0.0] * D
    v[0] = 1.0
    
    # Power iteration loop
    for k in range(N):
        if k % 2 == 0:
            # k=0, 2, 4, ... (Even iteration index -> A * v)
            # v_new = A * v_old
            v_next = mat_vec_mult_A(v)
        else:
            # k=1, 3, 5, ... (Odd iteration index -> A^T * v)
            # v_new = A^T * v_old
            v_next = mat_vec_mult_AT(v)
        
        # Normalize the resulting vector
        v = normalize(v_next)

    # The estimate of the spectral norm is the L2 norm of the final vector v_N.
    # Since v is already normalized at the end of the loop, its norm is 1.0.
    # However, the benchmark usually expects the norm *before* the final normalization,
    # or the norm of the vector resulting from the last multiplication step.
    # We re-calculate the norm of the final vector 'v' (which is already normalized)
    # to match the expected output format, which is usually the magnitude estimate.
    
    # If the power method estimate is the norm of the vector *before* normalization,
    # we need to track the unnormalized result. Let's re-run the last step without normalizing
    # to get the magnitude estimate, or simply use the norm of the last computed vector.
    
    # Since the loop structure above normalizes v at the end of every step, 
    # the final 'v' has norm 1.0. We must assume the benchmark expects the norm 
    # of the vector *before* the final normalization step, or perhaps the norm of the 
    # vector resulting from the last multiplication *before* normalization.
    
    # Let's re-run the last step to capture the unnormalized result magnitude.
    if N == 0:
        estimate = 1.0 # Or whatever the initial norm is
    elif N %
